### PR TITLE
[EC-19] Update Organization Settings Page

### PR DESCRIPTION
**Pull request by shane-melton**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Makes the following updates to the organization admin settings tab:

- Add the Policies page to the Settings tab
- Add the Single Sign-On page to the Settings tab
- Move the SSO identifier from Organization Info page and move it to the Single Sign-On page
  -  Includes a change server API model, see https://github.com/bitwarden/server/pull/2184
- Add new "enforcement" language to the organization Two-step Login page
- Update Single Single-On config page to use Component Library when available (check boxes and radio options still need to be updated once available)

### Related Refactoring
- Refactor the organization settings and policy folder to use NgModules and barrel files.
- Lazy load the new org settings module in the org routing module.

## Code changes

- **apps/web/src/app/organizations/settings/account.component.html**: Remove the SSO identifier from the org info page (it's still kept in the request model to support backwards compatibility) 

- **apps/web/src/app/organizations/settings/settings.component:** Add the new side nav options for SSO and Policies

- **apps/web/src/app/settings/two-factor-setup.component.html:** Adjust the language for organization two-step login enforcement  

- **bitwarden_license/bit-web/src/app/organizations/manage/sso.component:** Add SSO identifier form field and update to use CL

- **libs/common/src/models/request/organization/organizationSsoRequest.ts** Add a new `identifier` property for setting the SSO identifier from the SSO config page

- **libs/common/src/models/response/organization/organizationSsoResponse.ts:** Add and parse the new `identifier` property from the API 

### Refactors
- **apps/web/src/app/app.component.ts:** Policies are now in a barrel file and can be imported via a single path

- **apps/web/src/app/modules/loose-components.module.ts:** Remove the org settings and policy components from loose components module as they are now exported in their own modules

- **apps/web/src/app/organizations/settings/organization-settings.module.ts:** Created a settings NgModule and moved the org settings routing to its own module for lazy loading

- **apps/web/src/app/organizations/policies/policies.module.ts:** Add a policies NgModule 

- **apps/web/src/app/modules/organizations/billing/image-subscription-hidden.component**: Moved this component out of settings as it was only used in billing

### Other
- **libs/common/src/models/api/ssoConfigApi.ts:** Changed `spMinIncomingSigningAlgorithm` to a string to match server model
  - Same for **libs/common/src/models/view/ssoConfigView.ts**
 
- **libs/components/src/form-field/error.component.ts:** Added support for `maxLength` validator error message for CL inputs 

- **bitwarden_license/bit-web/src/app/organizations/components:** Remove obsolete input components after updating SSO page to Component Library

- **libs/angular/src/directives/copy-click.directive.ts:** Helper directive to easily copy a value to the clipboard on click.

- **libs/angular/src/directives/launch-click.directive.ts:** Helper directive to easily launch an external URI on click.

### Organization Info
image

### Policies
image

### Two-step Login
image

### Sigle Sign-On
image

## Before you submit

```
- [X] I have checked for **linting** errors (`npm run lint`) (required)
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)
```

### CODE REVIEW

File: apps/web/src/app/organizations/organization-routing.module.ts [hunk 1]
```diff
@@ -30,18 +27,7 @@ const routes: Routes = [
...
-        children: [
-          { path: "", pathMatch: "full", redirectTo: "account" },
-          { path: "account", component: AccountComponent, data: { titleId: "organizationInfo" } },
-          {
-            path: "two-factor",
-            component: TwoFactorSetupComponent,
-            data: { titleId: "twoStepLogin" },
-          },
-        ],
+        loadChildren: () => import("./settings").then((m) => m.OrganizationSettingsModule),
```
Comment (vincentsalucci) [inline apps/web/src/app/organizations/organization-routing.module.ts:30]: Ooo - this is new to me. Can you ELI5 what's going on?

File: libs/components/src/input/input.directive.ts [hunk 2]
```diff
@@ -1,6 +1,8 @@
 import { Directive, HostBinding, Input, Optional, Self } from "@angular/core";
 import { NgControl, Validators } from "@angular/forms";
 
+import { dirtyRequired } from "@bitwarden/angular/validators/dirty.validator";
```
Comment (vincentsalucci) [inline libs/components/src/input/input.directive.ts:4]: I might've missed it in the description summary, but what is this change for?

File: libs/angular/src/directives/launch-click.directive.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+@Directive({
+  selector: "[appLaunchClick]",
+})
+export class LaunchClickDirective {
+  constructor(private platformUtilsService: PlatformUtilsService) {}
+
+  @Input("appLaunchClick") uriToLaunch = "";
+
+  @HostListener("click") onClick() {
+    if (this.uriToLaunch) {
```
Comment (vincentsalucci) [inline libs/angular/src/directives/launch-click.directive.ts:14]: Would prefer an explicit conditional check here.

File: libs/angular/src/jslib.module.ts [hunk 4]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable import/order */
```
Comment (vincentsalucci) [inline libs/angular/src/jslib.module.ts:1]: What's up here?

File: bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts [hunk 5]
```diff
@@ -89,6 +89,8 @@ export class SsoComponent implements OnInit {
 
   enabled = this.formBuilder.control(false);
 
+  ssoIdentifier = this.formBuilder.control("", { validators: [Validators.maxLength(50)] });
```
Comment (vincentsalucci) [inline bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts:92]: @Hinton @MGibson1 Can either of you go through this class and take a look at the changes made, please.

File: apps/web/src/app/organizations/organization-routing.module.ts (hunk 1 above)
Comment (shane-melton) [reply apps/web/src/app/organizations/organization-routing.module.ts:30]: This allows Angular to [lazy load](https://angular.io/guide/lazy-loading-ngmodules) the org settings module only when the `/organizaitons/{ordId}/settings` route is loaded to save on the main bundle size. Any routes defined in the `OrganizationSettingsModule` will be registered as children of the `.../settings`.

It's also using the browser's[ native dynamic import functionality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import) instead of Angular's in-house version that was [deprecated in Angular v8 ](https://angular.io/guide/deprecations#loadchildren-string-syntax) since its now supported in most browsers.

File: libs/components/src/input/input.directive.ts (hunk 2 above)
Comment (shane-melton) [reply libs/components/src/input/input.directive.ts:4]: A `dirtyRequired` validator was being used in the SSO config form as a lot of the fields are only required in specific circumstances and are hidden otherwise. However, after switching the form over to the CL inputs, it only checked for Angular's built-in `Validators.required` validator when rendering the "Required" label on the form group. 

So this change includes checking if either the `dirtyRequired` **_or_** `Validators.required` validators are present before rendering that "Required" label.

File: libs/angular/src/jslib.module.ts (hunk 4 above)
Comment (shane-melton) [reply libs/angular/src/jslib.module.ts:1]: Oh good catch, looks like VS code disabled the eslint warning instead of auto fixing it.

File: bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts (hunk 5 above)
Comment (Hinton) [reply bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts:92]: We're now using deprecated overload of `group` in the `formBuilder`, if we want to migrate to using strongly typed forms (which we definitely do) we also need to update the calls to not call the deprecated https://angular.io/api/forms/FormBuilder#group

I also suspect we should combine these controls into a single form, but that's probably outside the scope of this task.

File: bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts [hunk 6]
```diff
@@ -204,6 +207,7 @@ export class SsoComponent implements OnInit {
 
     const request = new OrganizationSsoRequest();
     request.enabled = this.enabled.value;
+    request.identifier = this.ssoIdentifier.value;
     request.data = SsoConfigApi.fromView(this.ssoConfigForm.value as SsoConfigView);
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts:211]: I don't believe we should cast this after migrating to stronly typed forms.

File: bitwarden_license/bit-web/src/app/organizations/organizations.module.ts [hunk 7]
```diff
@@ -14,13 +11,6 @@ import { OrganizationsRoutingModule } from "./organizations-routing.module";
...
   imports: [SharedModule, OrganizationsRoutingModule],
-  declarations: [
-    InputCheckboxComponent,
-    InputTextComponent,
-    InputTextReadOnlyComponent,
-    SelectComponent,
-    SsoComponent,
-    ScimComponent,
-  ],
+  declarations: [InputCheckboxComponent, SsoComponent, ScimComponent],
```
[the remark this answers is not in the document]
Comment (shane-melton) [reply bitwarden_license/bit-web/src/app/organizations/organizations.module.ts:14]: Agreed, as we don't currently have a checkbox input component for the component library. But yeah, I would also suggest it be its own task.

File: apps/web/src/app/organizations/settings/organization-settings-routing.module.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+      },
+      {
+        path: "policies",
+        component: PoliciesComponent,
+        canActivate: [OrganizationPermissionsGuard],
+        data: {
+          organizationPermissions: (org: Organization) => org.canManagePolicies,
+          titleId: "policies",
+        },
+      },
```
[the remark this answers is not in the document]
Comment (shane-melton) [reply apps/web/src/app/organizations/settings/organization-settings-routing.module.ts:36]: We can definitely move the `PoliciesComponent` and `PolicyEditComponent` to the policy module. However, I originally thought it might be a bit verbose/overkill to move the single policy route to its own new `PolicyRoutingModule`. Of course, if we wanted to enable lazy loading for the policy module/page, then it would be worth it.

File: libs/components/src/input/input.directive.ts (hunk 2 above)
Comment (shane-melton) [reply libs/components/src/input/input.directive.ts:4]: It does look like we can use the `required` attribute directly (or the `Validators.required` in the form control) without issue. The one visual difference I see is the `bit-error-summary` component will "_prematurely_" show an error count for clean/empty form fields that are required. However, if this is the desired behavior for the error summary then we should be good to go with removing `dirtyRequired`. Just wanted to bring up this difference..

image

File: bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts (hunk 5 above)
Comment (shane-melton) [reply bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts:92]: Would you happen to have an example of what your envisioning for the strongly typed forms to look like? 

From what I can tell the `group()` overload being used is already the strongly typed one (even though the type is being inferred by the passed `controls` object). Should we be explicit with the generic type for the `group()` method, something like `group<SsoConfigView>(...)` instead? I found a potential method of doing this [here.](https://betterprogramming.pub/how-to-build-a-strongly-typed-angular-14-super-form-86837965a0e5)

I also see we could instead be using the `new FormControl()` for the form fields instead of the array syntax? e.g. going from `configType: [SsoType.None]` to `configType: new FormControl(SsoType.None)` Is this what you had in mind?

File: libs/components/src/input/input.directive.ts (hunk 2 above)
Comment (shane-melton) [reply libs/components/src/input/input.directive.ts]: I was able to remove the `dirtyRequired` validator and tweaked the error summary to stay in sync with the inline error messages.

File: bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts (hunk 6 above)
Comment (shane-melton) [reply bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts]: The cast is no longer required if using `getRawValue()` as it explicitly returns the strongly typed form type . `value` uses a `Partial<T>` of the form type, which required the cast.

**Comment by shane-melton** [comment]
@Hinton After our discussion regarding the `SsoConfigView` and `SsoConfigApi` types and some redundancy with the typed forms, I was able to re-use the `SsoConfigView` as an explicit typed form group using a new `ControlsOf<T>` type. ([Some more info here](https://netbasal.com/typed-reactive-forms-in-angular-no-longer-a-type-dream-bf6982b0af28))

It allows using a data model/interface as a strict type when defining form groups and provides type checking (and complains if a required data field is missing from the form group during its declaration). This effectively removes the redundant form group type. 

If we still want to remove/refactor the `SsoConfigView` and `SsoConfigApi` models we can do so, but I think it would be better left as its own task/ticket.

File: apps/web/src/app/organizations/organization-routing.module.ts [hunk 9]
```diff
@@ -34,18 +30,7 @@ const routes: Routes = [
...
```
Comment (Hinton) [inline apps/web/src/app/organizations/organization-routing.module.ts:33]: Have we done any analysis on the bundles to ensure lazy loading results in an improvement?

File: apps/web/src/app/organizations/settings/settings.component.ts [hunk 10]
```diff
@@ -11,20 +10,18 @@ import { PlatformUtilsService } from "@bitwarden/common/abstractions/platformUti
...
 
-  constructor(
-    private route: ActivatedRoute,
-    private organizationService: OrganizationService,
-    private platformUtilsService: PlatformUtilsService
-  ) {}
+  constructor(private route: ActivatedRoute, private organizationService: OrganizationService) {}
 
   ngOnInit() {
     // eslint-disable-next-line rxjs-angular/prefer-takeuntil, rxjs/no-async-subscribe
```
Comment (Hinton) [inline apps/web/src/app/organizations/settings/settings.component.ts:19]: Please remove the eslint comments now that we are modifying this file.

File: bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts [hunk 11]
```diff
@@ -111,7 +123,7 @@ export class SsoComponent implements OnInit {
     }
   );
 
-  samlForm = this.formBuilder.group(
+  private samlForm = this.formBuilder.group(
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts:126]: Shouldn't this also be using the ControlsOf helper?

File: bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts [hunk 12]
```diff
@@ -200,6 +221,7 @@ export class SsoComponent implements OnInit {
     this.validateForm(this.ssoConfigForm);
 
     if (this.ssoConfigForm.get("keyConnectorEnabled").value) {
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts:223]: Please use `ssoConfigForm.value.keyConnectorEnabled` wherever we use `get` since the tooling is better when doing refactors.

File: libs/components/src/form-field/error-summary.component.ts [hunk 13]
```diff
@@ -33,6 +33,10 @@ export class BitErrorSummary {
         return acc;
       }
 
+      if (!control.dirty && control.untouched) {
+        return acc;
+      }
```
Comment (Hinton) [inline libs/components/src/form-field/error-summary.component.ts:38]: Have we verified this will have no adversary effect on existing forms?

File: libs/angular/src/types/controls-of.ts [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+ * @example
+ * interface UserData {
+ *  name: string;
+ *  age: number;
+ * }
+ * const strictlyTypedForm = this.formBuilder.group<Controls<UserData>>({
+ *  name: new FormControl("John"),
+ *  age: new FormControl(24),
+ * });
+ */
```
Comment (Hinton) [inline libs/angular/src/types/controls-of.ts:16]: Should we provide credits for this? Source seems to be `https://netbasal.com/typed-reactive-forms-in-angular-no-longer-a-type-dream-bf6982b0af28`

File: apps/web/src/app/organizations/organization-routing.module.ts (hunk 9 above)
Comment (shane-melton) [reply apps/web/src/app/organizations/organization-routing.module.ts:33]: It will, however it does require tweaking the `DeleteOrganizationComponent` to _not_ be part of the org settings module (or it must be lazy loaded on its own). As its independently imported in the families for enterprise setup component which forces the settings module to be bundled in `main.js`. I can create a tech debt ticket to evaluate the best option to split that up to fully enable lazy loading.

With that said, I did a quick test to see the potential benefit, this will pull about **~10Kb** of gzipped JS out of the `main.js` bundle to be lazy loaded upon navigating to the org settings tab. 

image

File: bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts (hunk 11 above)
Comment (shane-melton) [reply bitwarden_license/bit-web/src/app/organizations/manage/sso.component.ts]: It didn't need it as it was already being strongly typed by the parent `ssoConfigForm` form group. However, I went ahead and added it, as it doesn't hurt to add the typing to be even more explicit. It also has the benefit of making it easier narrow down any future type errors should additional SSO fields be added to the form.

File: libs/angular/src/types/controls-of.ts (hunk 14 above)
Comment (shane-melton) [reply libs/angular/src/types/controls-of.ts:17]: Good point, added!

File: apps/web/src/app/organizations/settings/settings.component.ts (hunk 10 above)
Comment (shane-melton) [reply apps/web/src/app/organizations/settings/settings.component.ts]: Ah, missed this one after a merge from master, good catch, done!

File: libs/components/src/form-field/error-summary.component.ts (hunk 13 above)
Comment (shane-melton) [reply libs/components/src/form-field/error-summary.component.ts:38]: From what I can tell, the only other place the `<bit-error-summary>` component is used is in the register form at the moment. I verified that everything functioned as expected (in the register form's case, the error summary is manually hidden until the form is submitted anyways).

**Comment by shane-melton** [comment]
Thanks for taking another look at this @Hinton! I believe I've addressed all of your latest comments!

**Review by Hinton — APPROVED** [review]
Looks good!

File: apps/web/src/app/organizations/organization-routing.module.ts (hunk 9 above)
Comment (Hinton) [reply apps/web/src/app/organizations/organization-routing.module.ts:33]: In that case maybe we shouldn't lazy load it. 10Kb is not much.
